akka-http-caching/src/main/scala/akka/http/caching/scaladsl/Cache.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com/>
+ */
+package akka.http.caching.scaladsl
+
+import java.util.Optional
+import java.util.concurrent.{ CompletableFuture, CompletionStage }
+
+import akka.annotation.{ ApiMayChange, DoNotInherit }
+import akka.japi.{ Creator, Procedure }
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable
+import scala.compat.java8.FutureConverters.{ toJava ⇒ futureToJava, toScala ⇒ futureToScala }
+import scala.concurrent.{ Future, Promise }
+
+/**
+ * API MAY CHANGE
+ *
+ * General interface implemented by all akka-http cache implementations.
+ */
+@ApiMayChange
+@DoNotInherit
+abstract class Cache[K, V] extends akka.http.caching.javadsl.Cache[K, V] {
+  cache ⇒
+
+  /**
+   * Returns either the cached Future for the given key or evaluates the given value generating
+   * function producing a `Future[V]`.
+   */
+  def apply(key: K, genValue: () ⇒ Future[V]): Future[V]
+
+  /**
+   * Returns either the cached Future for the key or evaluates the given function which
+   * should lead to eventual completion of the promise.
+   */
+  def apply(key: K, f: Promise[V] ⇒ Unit): Future[V] =
+    apply(key, () ⇒ { val p = Promise[V](); f(p); p.future })
+
+  /**
+   * Returns either the cached Future for the given key or the given value as a Future
+   */
+  def get(key: K, block: () ⇒ V): Future[V] =
```

Comment:
Note that `block` is not GC friendly, since it requires allocating a capturing lambda for the key. This is a small overhead, but worth being aware of. The alternative, e.g. in `Map#computeIfAbsent`, is to pass a function `k -> v` so that the caller can opt into the lamda being cached by the JVM.

Reply:
People can store the function though -- `val f: () => V = () => "hello"; get(key, f)`

Reply:
Sure, but what about common usages like,

```
cache.get(userId, () -> repository.getById(userId))
```

The lambda has to be allocated to pass the key through.

Reply:
Adding the `get(K, V)` overload is not a problem, we can do it after merging this.
It's still possible to change APIs, and this PR has been running since March.

I agree and appropriate the feedback a lot, but also would want this *initial* (!) PR be finally merged

Reply:
How about we open tickets for these polishing-up tasks? Pretty sure people will happily pick them up?

Reply:
Sure, any observations I make should not be considered blockers and just fyi.

Reply:
Agreed, will make them tickets 👍 
Only making sure we don't open the nitpicking for a week jar again without merging like has happened a few times (so team or anyone would not be sure if "to merge or not yet") :-) Please do keep the feedback coming, will ticketify and follow up!